dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java

```diff
@@ -0,0 +1,455 @@
+ */
+@Component(configurationPolicy = ConfigurationPolicy.IGNORE)
+public class ModuleSelectionConfigImpl implements ModuleSelectionConfig, OpenAPIAppConfigProvider.OpenAPIAppConfigListener {
+
+    private static final TraceComponent tc = Tr.register(ModuleSelectionConfigImpl.class);
+
+    private volatile ConfigValues configValues = null;
+
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.STATIC, unbind = "unbindAppConfigListener")
+    private OpenAPIAppConfigProvider configFromServerXML;
+
+    @Activate
+    public void activate() {
+        configFromServerXML.registerAppConfigListener(this);
+    }
+
+    @Deactivate
+    public void deactivate() {
+        configFromServerXML.unregisterAppConfigListener(this);
+    }
+
+    @Override
+    public String toString() {
+        if (configValues == null) {
+            return ("Unconfigured Module Selection Config");
+        }
+
+        StringBuilder sb = new StringBuilder("Module Selection Config[");
+        if (getConfigValues().isFirst) {
+            sb.append("useFirstModuleOnly");
+        } else {
+            if (getConfigValues().isAll) {
+                sb.append("include = all");
+            } else {
+                sb.append("include = ").append(getConfigValues().included);
+            }
+            sb.append(", ");
+            sb.append("exclude = ").append(getConfigValues().excluded);
+        }
+        sb.append("]");
+        return sb.toString();
+    }
+
+    /**
+     * Whether the legacy "first module only" mode should be used.
+     * <p>
+     * As this requires special handling, if this method returns {@code true}, the other methods on this object should not be called.
+     *
+     * @return {@code true} if only the first module found should be processed for OpenAPI annotations, {@code false} otherwise.
+     */
+    @Override
+    public boolean useFirstModuleOnly() {
+        return getConfigValues().isFirst;
+    }
+
+    /**
+     * Whether the given module should be used to create the OpenAPI document, based on the config
+     *
+     * @param module the module to check
+     * @return {@code true} if the module should be used, {@code false} otherwise
+     */
+    @Override
+    public boolean isIncluded(ModuleInfo module) {
+        if (getConfigValues().isFirst) {
+            return true;
+        }
+
+        boolean result = false;
+        if (getConfigValues().isAll) {
+            result = true;
+        } else {
+            for (ModuleName name : getConfigValues().included) {
+                if (matches(name, module)) {
+                    result = true;
+                    break;
+                }
+            }
+        }
+
+        if (result) {
+            for (ModuleName name : getConfigValues().excluded) {
+                if (matches(name, module)) {
+                    result = false;
+                    break;
+                }
+            }
+        }
+
+        return result;
+    }
```

Comment:
We have a problem with consistency in many of the methods in this class.

Most methods rely on having a consistent view of the config that doesn't change throughout the method. In that case, they need to call `getConfigValues()` only once and re-use the result.